src/checker.rs

```diff
@@ -15,62 +21,153 @@ pub enum FailureReason {
     DnsError(String),
     Error(String),
 }
+
 /// Fetches the response from a URL.
 ///
 /// First attempts to fetch just the head, and if not supported falls back to fetching the entire body.
-async fn do_request(client: &reqwest::Client, url: &str) -> Result<Response, reqwest::Error> {
-    let response = client.head(url).send().await?;
-    match response.status() {
-        StatusCode::METHOD_NOT_ALLOWED => client.get(url).send().await,
-        _ => Ok(response),
+async fn do_request(
+    client: &reqwest::Client,
+    url: &str,
+) -> (RequestType, Result<Response, reqwest::Error>) {
+    let head_response = match client.head(url).send().await {
+        Ok(response) => response,
+        Err(e) => return (RequestType::Head, Err(e)),
+    };
+
+    // Successful head request
+    if head_response.status() != StatusCode::METHOD_NOT_ALLOWED {
+        return (RequestType::Head, Ok(head_response));
+    }
+
+    let result = client.get(url).send().await;
+    (RequestType::Get, result)
+}
+
+fn dns_error(err: &reqwest::Error) -> Option<String> {
+    let mut inner = &err as &dyn Error;
+    while let Some(source) = inner.source() {
+        inner = source;
+
+        // TODO: Would be better to get specific errors without string matching like this
+        // Not sure if there's a better way
+        let inner_message = inner.to_string();
+        if inner_message.contains("dns error") {
+            return Some(inner.source().unwrap().to_string());
+        }
     }
+    None
 }
 
 /// Makes a request to a url to determine whether it is up.
 /// Up is defined as returning a 2xx within a specific timeframe.
 pub async fn check_url(client: &reqwest::Client, url: &str) -> CheckResult {
-    match do_request(client, url).await {
-        Ok(_) => CheckResult::Success,
-        Err(e) => {
+    let trace_id = Uuid::new_v4();
+
+    let start = Instant::now();
+    let (request_type, response) = do_request(client, url).await;
+    let duration_ms = Some(start.elapsed().as_millis());
+
+    let status = match response.as_ref().map_or(false, |r| r.status().is_success()) {
+        true => CheckStatus::Success,
+        false => CheckStatus::Failure,
+    };
+
+    let http_status_code = response
+        .as_ref()
+        .map(|r| Some(r.status()))
+        .unwrap_or_else(|e| e.status())
+        .map(|s| s.as_u16());
+
+    let request_info = Some(RequestInfo {
+        http_status_code,
+        request_type,
+    });
+
+    let status_reason = match response {
+        Ok(r) => match r.status().is_success() {
+            true => None,
+            false => Some(CheckStatusReason {
+                status_type: CheckStatusReasonType::Failure,
+                description: format!("Got non 2xx status: {}", r.status()),
+            }),
+        },
```

Comment:
Here I would suggest a match guard:
```suggestion
        Ok(r) if !r.status().is_success() => Some(CheckStatusReason {
            status_type: CheckStatusReasonType::Failure,
            description: format!("Got non 2xx status: {}", r.status()),
        }),
        Ok(_) => None,
```
Or alternatively
```suggestion
        Ok(r) if r.status().is_success() => None,
        Ok(r) => Some(CheckStatusReason {
            status_type: CheckStatusReasonType::Failure,
            description: format!("Got non 2xx status: {}", r.status()),
        }),
```
if you don't want to have the negated conditional.

Reply:
Ohhh that's clean

Reply:
I didn't know about match guards thank you